server/src/main/java/io/druid/server/log/LoggingRequestLogger.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.log;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.metamx.common.logger.Logger;
+import io.druid.query.Query;
+import io.druid.server.RequestLogLine;
+import org.slf4j.MDC;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class LoggingRequestLogger implements RequestLogger
+{
+  final ObjectMapper mapper;
+  final boolean setMDC;
+  final boolean setContextMDC;
+
+  public LoggingRequestLogger(
+      ObjectMapper mapper,
+      boolean setMDC,
+      boolean setContextMDC
+  )
+  {
+    this.mapper = mapper;
+    this.setMDC = setMDC;
+    this.setContextMDC = setContextMDC;
+  }
+
+  private static final Logger LOG = new Logger(LoggingRequestLogger.class);
+
+  @Override
+  public void log(RequestLogLine requestLogLine) throws IOException
+  {
+    final Map mdc;
+    if (setMDC) {
+      mdc = MDC.getCopyOfContextMap();
+    } else {
+      mdc = null;
+    }
+    try {
+      if (setMDC) {
+        try {
+          final Query query = requestLogLine.getQuery();
+          MDC.put("queryId", query.getId());
+          MDC.put("dataSource", query.getDataSource().toString());
+          MDC.put("queryType", query.getType());
+          MDC.put("hasFilters", Boolean.toString(query.hasFilters()));
+          MDC.put("remoteAddr", requestLogLine.getRemoteAddr());
+          MDC.put("duration", query.getDuration().toString());
+          MDC.put("resultOrdering", query.getResultOrdering().toString());
+          MDC.put("descending", Boolean.toString(query.isDescending()));
+          if (setContextMDC) {
+            final Iterable<Map.Entry<String, Object>> entries = query.getContext() == null
+                                                                ? ImmutableList.<Map.Entry<String, Object>>of()
+                                                                : query.getContext().entrySet();
+            for (Map.Entry<String, Object> entry : entries) {
+              MDC.put(entry.getKey(), entry.getValue() == null ? "NULL" : entry.getValue().toString());
+            }
+          }
+        }
+        catch (RuntimeException re) {
+          LOG.error(re, "Error preparing MDC");
+        }
+      }
+      final String line = requestLogLine.getLine(mapper);
+      LOG.info("%s", line);
+    }
+    finally {
+      if (setMDC && mdc != null) {
+        MDC.setContextMap(mdc);
+      }
+    }
+  }
+
+  public boolean isSetMDC()
+  {
+    return setMDC;
+  }
+
+  public boolean isSetContextMDC()
```

Comment:
YAGNI and minimizing API surface area are good enough reasons for me, but not enough to make a big deal about it. So I'm cool with this the way it is.